src/storage-types/src/connections/aws.rs

```diff
@@ -136,58 +227,119 @@ impl AwsConfig {
-                None => None,
-            },
-        ));
-
-        if let Some(AwsAssumeRole { arn }) = &self.role {
-            let mut role = AssumeRoleProvider::builder(arn).session_name("materialize");
-            // This affects which region to perform STS on, not where
-            // anything else happens.
-            if let Some(region) = &region {
-                role = role.region(region.clone());
+        let cred_provider = match &self.auth {
+            AwsAuth::Credentials(credentials) => {
+                let AwsCredentials {
+                    access_key_id,
+                    secret_access_key,
+                    session_token,
+                } = credentials;
+
+                SharedCredentialsProvider::new(Credentials::from_keys(
+                    access_key_id.get_string(secrets_reader).await.unwrap(),
+                    secrets_reader
+                        .read_string(*secret_access_key)
+                        .await
+                        .unwrap(),
+                    match session_token {
+                        Some(t) => Some(t.get_string(secrets_reader).await.unwrap()),
+                        None => None,
+                    },
+                ))
             }
-            if let Some(external_id_prefix) = external_id_prefix {
-                let external_id = if let Some(suffix) = external_id_suffix {
-                    format!("{}-{}", external_id_prefix, suffix)
-                } else {
-                    external_id_prefix.to_string()
-                };
-                role = role.external_id(external_id);
+            AwsAuth::AssumeRole(assume_role) => {
+                let sts_client = aws_sdk_sts::Client::new(&aws_config::from_env().load().await);
+                let AwsAssumeRole {
+                    arn,
+                    session_name,
+                    mz_principal,
+                } = assume_role;
+                let assume_role_response = sts_client
+                    .assume_role()
+                    .role_arn(mz_principal.arn.clone())
+                    .role_session_name("materialize_assume_role")
+                    .send()
+                    .await?;
+
+                let temp_credentials = assume_role_response.credentials().ok_or_else(|| {
+                    anyhow!("Could not fetch temporary credentials with materialize role")
+                })?;
+
+                let access_key = temp_credentials
+                    .access_key_id()
+                    .ok_or_else(|| anyhow!("missing access_key"))?;
+                let secret_key = temp_credentials
+                    .secret_access_key()
+                    .ok_or_else(|| anyhow!("missing secret key"))?;
+
+                let credentials = aws_credential_types::Credentials::from_keys(
+                    access_key,
+                    secret_key,
+                    temp_credentials.session_token().map(|t| t.to_string()),
+                );
```

Comment:
I'm not sure this is the right way to do role chaining. Unfortunately I can't find an example online, but I think you want to stack two `AssumeRoleProvider`s together somehow. Otherwise, when the credentials that got "hardcoded" into `credentials` expire, the SDK credentials will be permanently expired—instead of the `AssumeRoleProvider` just fetching new credentials. I think!

Reply:
So, I saw the following in the docs https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_terms-and-concepts.html
> Role chaining is when you use a role to assume a second role through the AWS CLI or API. For example, RoleA has permission to assume RoleB. You can enable User1 to assume RoleA by using their long-term user credentials in the AssumeRole API operation. This returns RoleA **short-term credentials**. With role chaining, you can use RoleA's short-term credentials to enable User1 to assume RoleB.

Let me double check the API. I found a java example https://exceptionly.com/2022/03/09/aws-iam-role-chaining-in-java-dynamodb-s3/ which does something similar to what you mentioned. I think rust should be similar.

Reply:
(I was not able to test out the assume role flow since the new role changes were not in place)